### PR TITLE
Add module builder

### LINKED ISSUE #22: PR to support modular interface?
Hello, and thanks for this neat and useful toolchain.  I'm trying to do a better job of managing my OpenWRT-based router and APs, and as a confirmed Nixaholic this repository is just what I needed.

Would you be interested in a PR introducing a modular interface to the OpenWRT image builder code?  I have a (very) rough cut at such an interface up on SourceHut [here](https://git.sr.ht/~tomeon/nix-openwrt-imagebuilder-module).

The high-level view:

```nix
{
  openwrtImages.default = self.lib.openwrtImage {
    modules = [
      {
        image = {
          profileName = "avm_fritz7412";
          packages.include = ["tcpdump"];
          packages.exclude = ["ppp"];
          disabledServices = ["dnsmasq"];
          snippets."etc/uci-defaults/99-custom".text = ''
            uci -q batch << EOI
            set system.@system[0].hostname='testap'
            commit
            EOI
          '';
        };
      }
    ];
  };
}
```

If specified as a flake output, the image and associated assets could then be built with `nix build '.#openwrtImages.default.config.system.build.image' `.  I plan to add a wrapper command for this that would permit doing, say, `nix-openwrt-imagebuild '.#default'`.

Any interest in integrating something like this in your project?

Thank you!

### PR DESCRIPTION
**Pull request by pedorich-n**
This PR adds a module interface to the builder derivation. This brings benefits such as strict options definition, better type-checks, and flexibility for future improvements or modifications. 
I left the old builder in place for backward compatibility. In the future, it could be removed, and an "adapter" could be installed to convert old inputs into the new module system and maintain backward compatibility.

The module is inspired by @tomeon's [nix-openwrt-imagebuilder-module](https://git.sr.ht/~tomeon/nix-openwrt-imagebuilder-module/tree/6c2a5fc19c5a6fd532b99c69e5b4ce14adfd80fe/item/module.nix) project, but it's structured differently and the module itself is simpler. Please let me know if you feel it isn't fair to your work. 

Opening this as a draft to gather opinions. 
If everything is fine I will add module documentation and update README to tell users to use `build-module` in a follow-up PR to not pollute this one. 

I also moved examples into a dedicated folder, since we have 4 of them now.

Closes #22

### CODE REVIEW

File: builder/module-options.nix [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        example = false;
+        description = ''
+          Store locally generated signing key in built images
+
+          Specify `null` if you want to use default from OpenWRT imagebuilder
+        '';
+      };
+    };
+
+    sumsFileSha256 = lib.mkOption {
```
Comment (pedorich-n) [inline builder/module-options.nix:207]: It's hard to come up with a good name for this one 😅

File: builder/module-options.nix [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+  hashedTarget = releaseHashes.targets.${config.hardware.target}.${config.hardware.subtarget};
+  hashedFeeds = releaseHashes.packages.${config.hardware.arch};
+
+  defaultSumsFileSha256 = hashedTarget.sha256;
+
+  # TODO: get rid of `pkgs`
+  profiles = import ../profiles.nix { pkgs = null; inherit (config) release; };
+  detectedProfile = profiles.identifyProfile config.hardware.profile;
+  # TODO: move arch to profiles
+  detectedArch = hashedTarget.packagesArch;
```
Comment (pedorich-n) [inline builder/module-options.nix:16]: `arch` is already a property in `profiles.json`. I feel like it's more fitting to get this info from there, than from the feeds, since it's a property of a specific board/target

**Review by pedorich-n — COMMENTED** [review]
Also, I would love for people to test this, maybe I missed some corner cases.
In my tests, I tried building with packages, folders, and individual files. I haven't tried disabling services, removing packages or anything else.

**Comment by astro** [comment]
Cc: @ehmry @adamcstephens @makisekurisu

**Comment by ehmry** [comment]
I'd like to review but I haven't been using the imagebuilder.

File: builder/module-options.nix (hunk 1 above)
Comment (SuperSandro2000) [reply builder/module-options.nix:207]: We probably want to use Hash instead of Sha256 for future proofness.

File: builder/module-config.nix [hunk 3]
```diff
@@ -0,0 +1,169 @@
+{ config, pkgs, lib, ... }:
+let
+  inherit (import ../files.nix {
+    inherit pkgs;
+    inherit (config) release feedsSha256;
+    inherit (config.hardware) target;
+    variant = config.hardware.subtarget;
```
Comment (SuperSandro2000) [inline builder/module-config.nix:7]: Shouldn't we name variant subtarget here, too or at least add a TODO comment to rename it?

File: builder/module-config.nix [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    packagesArch = config.hardware.arch;
+    sha256 = config.sumsFileSha256;
```
Comment (SuperSandro2000) [inline builder/module-config.nix:9]:
```suggestion
    hash = config.sumsFileHash;
```

File: builder/module-config.nix [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    profiles.profiles.${config.hardware.profile}.device_packages or [ ] ++
+    customPackagesFormatted;
+
+  allRequiredPackageFiles = expandDeps allPackages requiredPackages;
+
+  mkPackageSymlink = pname:
+    let
+      package = allPackages.${pname};
+    in
+    lib.optional (package.type == "real") ''[ -e "packages/${package.filename}" ] || ln -s ${package.file} "packages/${package.filename}"'';
```
Comment (SuperSandro2000) [inline builder/module-config.nix:26]:
```suggestion
    lib.optional (package.type == "real") ''
      [ -e "packages/${package.filename}" ] || ln -s ${package.file} "packages/${package.filename}"
    '';
```

File: builder/module-config.nix [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  );
+
+  mkCopyStatement = file: ''(
+     target_rel=${lib.escapeShellArg file.target}
+     target="files/''${target_rel#/}"
+     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
+        mkdir -p $(dirname "''${target}")
+     ''}
+     cp --recursive --no-preserve=all "${file.source}" "''${target}"
+    )'';
```
Comment (SuperSandro2000) [inline builder/module-config.nix:41]:
```suggestion
  mkCopyStatement = file: ''
    (
     target_rel=${lib.escapeShellArg file.target}
     target="files/''${target_rel#/}"
     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
        mkdir -p "$(dirname "''${target}")"
     ''}
     cp -r --no-preserve=all "${file.source}" "''${target}"
    )
  '';
```

File: builder/module-config.nix [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  name = lib.concatStringsSep "-" (
+    [ "openwrt" config.release ] ++
+    lib.optional (config.image.extraName != null) config.image.extraName ++
+    [ config.hardware.target config.hardware.subtarget config.hardware.profile ]
+  );
+
+  src =
+    let
+      inherit (pkgs.stdenv.hostPlatform) uname;
+      imageBuilderPrefix = "openwrt-imagebuilder-${if config.release == "snapshot" then "" else "${config.release}-"}";
```
Comment (SuperSandro2000) [inline builder/module-config.nix:75]:
```suggestion
      imageBuilderPrefix = "openwrt-imagebuilder-${lib.optionalString (config.release != "snapshot") "${config.release}-"}";
```

File: builder/module-config.nix [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      baseFileName = "${imageBuilderPrefix}${config.hardware.target}-${config.hardware.subtarget}.${uname.system}-${uname.processor}";
+      possibleFileNames = builtins.map (extension: "${baseFileName}${extension}") [ ".tar.zst" ".tar.xz" ];
+      matches = builtins.filter (fileName: builtins.hasAttr fileName variantFiles) possibleFileNames;
+    in
+    if matches != [ ]
+    then builtins.getAttr (builtins.elemAt matches 0) variantFiles
+    else throw "No valid image builder found!";
```
Comment (SuperSandro2000) [inline builder/module-config.nix:82]: Can we easily name which one we searched for to make debugging easier?

File: builder/module-config.nix [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+in
+{
+
+  ###### implementation
```
Comment (SuperSandro2000) [inline builder/module-config.nix:86]:
```suggestion
```
lets no copy paste legacy comments from nixos

File: builder/module-docs.nix [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  # Inspired by https://github.com/astro/microvm.nix/blob/e8d5f12b834a59187c7ec147a8952a0567f97939/pkgs/doc.nix
+  extraConfig = { lib, ... }: {
+    options._module.args = lib.mkOption {
+      internal = true; # Hide `_module` from the docs
+    };
+
+    config._module = {
+      check = false; # Don't check for unset variables
+      args = {
+        inherit pkgs lib; # Provide pkgs and lib to modules
```
Comment (SuperSandro2000) [inline builder/module-docs.nix:12]: From my experience with nüschtos search you generally don't need to provide lib.

File: builder/module-docs.nix [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    (lib.evalModules {
+      modules = [
+        extraConfig
+        module
+      ];
+    }).options;
+
+  makeOptionsDoc = module: (pkgs.nixosOptionsDoc {
+    options = optionsFor module;
+    transformOptions = options: builtins.removeAttrs options [ "declarations" ];
```
Comment (SuperSandro2000) [inline builder/module-docs.nix:27]: We probably want to filter declarations separately and also provide the optionsJSON to easily integrate it into nüschtos search.

File: builder/module-options.nix [hunk 12]
```diff
@@ -0,0 +1,272 @@
+{ config, lib, ... }:
+# TODO: add support for assertions
+let
+  latestRelease = import ../latest-release.nix;
+
+  releaseHashes = import ../hashes/${config.release}.nix;
+  hashedTarget = releaseHashes.targets.${config.hardware.target}.${config.hardware.subtarget};
+  hashedFeeds = releaseHashes.packages.${config.hardware.arch};
```
Comment (SuperSandro2000) [inline builder/module-options.nix:8]: We should probably add a few or's here to make errors easier to debug.

File: builder/module-options.nix [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        example = lib.literalExpression "./uci-script";
+        description = ''
+          Path to the source file or folder.
+        '';
+      };
+    };
+  };
+in
+{
+  ###### interface
```
Comment (SuperSandro2000) [inline builder/module-options.nix:40]:
```suggestion
```

File: builder/module-config.nix (hunk 3 above)
Comment (pedorich-n) [reply builder/module-config.nix:7]: I didn't want this PR to turn into a whole repository rewrite. But I'll add a TODO.

File: builder/module-docs.nix (hunk 11 above)
Comment (pedorich-n) [reply builder/module-docs.nix:27]: I'm not familiar with "nüschtos search". Do you have an example of how the module should look like?

Comment (SuperSandro2000) [reply builder/module-docs.nix:24]: It should look similar to this https://github.com/NuschtOS/search/blob/3051be7f403bff1d1d380e4612f0c70675b44fc9/nix/wrapper.nix#L4-L17

**Comment by astro** [comment]
I like this. Please feel encouraged to follow through.

**Comment by pedorich-n** [comment]
Thanks for your encouragement, @astro !

I've updated the docs module, exposing both markdown and json. At a certain iteration, I realized it looked almost like https://github.com/astro/microvm.nix/blob/2ae08de8e8068b00193b9cfbc0acc9dfdda03181/pkgs/doc.nix, so I just copied most of the stuff from there 😅 
@SuperSandro2000 please have a look.

I also added support for "separate Kmods" and tested the build with both "old" (23.05.5) and "new" (24.10.0-rc4) versions. Builds fine.

**Comment by pedorich-n** [comment]
Happy New Year everybody!

I've added a new package to generate/update module docs and a GitHub flow to do it from GHA. 
I've also updated the documentation with a minimal example and a recommendation to use the builder module over the builder function.

I think this PR is ready for a proper review now.

**Comment by pedorich-n** [comment]
Is there any chance I could get a review @SuperSandro2000 @astro ? 👀

File: builder/module-config.nix [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        zlib
+        zstd
+      ] ++ lib.optional (lib.versionOlder config.release "21" && config.release != "snapshot") python2;
+
+      postPatch = ''
+        patchShebangs scripts staging_dir/host/bin
+        substituteInPlace rules.mk \
+          --replace "SHELL:=/usr/bin/env bash" "SHELL:=${pkgs.runtimeShell}" \
+          --replace "/usr/bin/env true" "${pkgs.coreutils}/bin/true" \
+          --replace "/usr/bin/env false" "${pkgs.coreutils}/bin/false"
```
Comment (SuperSandro2000) [inline builder/module-config.nix:123]: Should we use --replace-fail here?

File: builder/module-options.nix [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    };
+  };
+in
+{
+  # Inspired by https://git.sr.ht/~tomeon/nix-openwrt-imagebuilder-module/tree/6c2a5fc19c5a6fd532b99c69e5b4ce14adfd80fe/item/module.nix
+  options = {
+    release = lib.mkOption {
+      type = lib.types.nonEmptyStr;
+      default = latestRelease;
+      defaultText = lib.literalMD "Value of `latest-release.nix`";
```
Comment (SuperSandro2000) [inline builder/module-options.nix:56]:
```suggestion
      defaultText = "Value of `latest-release.nix`";
```
literalMD doesn't exist anymore in unstable

File: docs/module-options.md
Comment (SuperSandro2000) [inline docs/module-options.md:1]: I would not commit this file and generate it on the fly for github pages or use https://github.com/nuschtos/search

**Comment by SuperSandro2000** [comment]
I would really need to try this to know fore sure if it works perfectly but I am not keen on bricking my router right now in case of any issues 😓

File: builder/module-config.nix [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
+        mkdir -p "$(dirname "''${target}")"
+     ''}
+     cp -r --no-preserve=all "${file.source}" "''${target}"
+    )
+  '';
+
+  copyFiles = pkgs.writeScript "copy-openwrt-files" (
+    lib.concatLines (builtins.map mkCopyStatement config.files)
+  );
```
Comment (astro) [inline builder/module-config.nix:51]: I actually prefer having once-used definitions inline for readability. The indentation isn't very deep here so I think it would be fine:

```suggestion
  copyFiles = pkgs.writeScript "copy-openwrt-files" (
    lib.concatMapStrings (file: ''
      (
        target_rel=${lib.escapeShellArg file.target}
        target="files/''${target_rel#/}"
        ${lib.optionalString (lib.pathIsRegularFile file.source) ''
          mkdir -p "$(dirname "''${target}")"
        ''}
        cp -r --no-preserve=all "${file.source}" "''${target}"
      )
    '') config.files
  );
```

**Review by astro — CHANGES_REQUESTED** [review]
As you already said, if I merge the modules like this we're creating duplicate functionality in this project. Any intermediate PR would have to do twice the work.

That's why I am going to wait for the removal of the old code and the addition of a few shims.

**Comment by SuperSandro2000** [comment]
Me and @m4rc3l have some interest in using this for future infrastructure. Hopefully we find some time bringing this forward together.

**Comment by pedorich-n** [comment]
I'm sorry I haven't been active. I really want to see this working, but I struggled to find the willpower to get back to it.
The changes in this PR have grown quite large, and being asked to completely replace existing code within the same PR became overwhelming for me.

I'd still love to help if we can find a way to contribute incrementally.

**Comment by MarcelCoding** [comment]
I've successfully deployed my testing AP using this PR in https://gitea.c3d2.de/zentralwerk/network/src/branch/openwrt-module. (I've used my fork to get up-to-date hashes)
